File: src/tools/scope_serial.py
#
# Copyright (c) 2021-present LAAS-CNRS
#
# SPDX-License-Identifier: GPL-2.0-or-later
#

"""Bounded parser for the dedicated OwnTech scope-data serial interface."""

from dataclasses import dataclass
import math
import re
import struct
import time

import serial


SCOPE_BAUDRATE = 115200
SCOPE_SAMPLE_COUNT = 1024
SCOPE_CHANNEL_NAMES = (
    "V1Low_V",
    "V2Low_V",
    "VHigh_V",
    "I1Low_A",
    "I2Low_A",
    "IHigh_A",
    "Duty1",
    "Duty2",
)

_PROBE_RESPONSE = "SCOPE-DATA/1 OK"
_BEGIN_RECORD = "begin record"
_END_RECORD = "end record"
_FINAL_INDEX_RE = re.compile(r"# ([0-9]+)")
_HEX_VALUE_RE = re.compile(r"[0-9A-Fa-f]{8}")


class ScopeSerialError(RuntimeError):
    """Raised for scope transport, framing, validation, or timeout errors."""


@dataclass(frozen=True)
class ScopeCapture:
    """One decoded capture in chronological sample order."""

    decimation: int
    sample_period_us: int
    duration_ms: float
    channel_names: tuple
    samples: tuple
    final_index: int
    pretrigger_ratio: float
    time_axis_s: tuple


class ScopeSerial:
    """Read the fixed scope protocol from an explicitly selected data port."""

    def __init__(
        self,
        port,
        *,
        timeout=1.0,
        probe_timeout=3.0,
        serial_factory=None,
    ):
        if not isinstance(port, str) or not port:
            raise ValueError("scope port must be an explicit non-empty string")
        if isinstance(timeout, bool) or not isinstance(timeout, (int, float)):
            raise ValueError("timeout must be a positive finite number")
        timeout = float(timeout)
        if not math.isfinite(timeout) or timeout <= 0:
            raise ValueError("timeout must be a positive finite number")
        if (
            isinstance(probe_timeout, bool)
            or not isinstance(probe_timeout, (int, float))
        ):
            raise ValueError("probe_timeout must be a positive finite number")
        probe_timeout = float(probe_timeout)
        if not math.isfinite(probe_timeout) or probe_timeout <= 0:
            raise ValueError("probe_timeout must be a positive finite number")

        factory = serial.Serial if serial_factory is None else serial_factory
        self.port = port
        self.ser = None
        try:
            self.ser = factory(
                port=port,
                baudrate=SCOPE_BAUDRATE,
                timeout=timeout,
                write_timeout=timeout,
            )
            self._probe(probe_timeout)
        except Exception as exc:
            if self.ser is not None:
                self.ser.close()
            if isinstance(exc, (ScopeSerialError, ValueError)):
                raise
            raise ScopeSerialError(
                f"failed to open scope data port {port!r}: {exc}"
            ) from exc

    def close(self):
        if self.ser is not None:
            self.ser.close()

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc, traceback):
        self.close()

    def __repr__(self):
        return f"<ScopeSerial {self.port}>"

    def _write_command(self, command):
        if callable(getattr(self.ser, "reset_input_buffer", None)):
            self.ser.reset_input_buffer()
        try:
            written = self.ser.write(command)
            if written is not None and written != len(command):
                raise ScopeSerialError(
                    f"short write on scope data port: {written}/{len(command)}"
                )
            if callable(getattr(self.ser, "flush", None)):
                self.ser.flush()
        except ScopeSerialError:
            raise
        except Exception as exc:
            raise ScopeSerialError(
                f"scope data-port write failed: {exc}"
            ) from exc

    def _readline(self, deadline, *, maximum_length=256):
        data = bytearray()
        while time.monotonic() < deadline:
            try:
                chunk = self.ser.read(1)
            except Exception as exc:
                raise ScopeSerialError(
                    f"scope data-port read failed: {exc}"
                ) from exc
            if not chunk:
                continue
            data.extend(chunk)
            if len(data) > maximum_length:
                raise ScopeSerialError(
                    f"scope line exceeds {maximum_length} bytes"
                )
            if chunk == b"\n":
                try:
                    return data.rstrip(b"\r\n").decode("ascii")
                except UnicodeDecodeError as exc:
                    raise ScopeSerialError(
                        "scope response is not ASCII"
                    ) from exc
        raise ScopeSerialError("scope data-port response timed out")

    def _probe(self, timeout):
        self._write_command(b"?")
        deadline = time.monotonic() + timeout
        for _ in range(32):
            line = self._readline(deadline)
            if line == _PROBE_RESPONSE:
                return
            if line.startswith("SCOPE-DATA/1 ERROR"):
                raise ScopeSerialError(f"scope probe rejected: {line}")
        raise ScopeSerialError(
            f"scope probe mismatch: {_PROBE_RESPONSE!r} was not received"
        )

    @staticmethod
    def _validate_capture_metadata(
        decimation,
        sample_period_us,
        duration_ms,
        pretrigger_ratio,
    ):
        if (
            isinstance(decimation, bool)
            or not isinstance(decimation, int)
            or not 1 <= decimation <= 100
        ):
            raise ValueError("decimation must be an integer between 1 and 100")
        if (
            isinstance(sample_period_us, bool)
            or not isinstance(sample_period_us, int)
            or sample_period_us != 100 * decimation
        ):
            raise ValueError(
                "sample_period_us must equal 100 times decimation"
            )
        if isinstance(duration_ms, bool) or not isinstance(
            duration_ms, (int, float)
        ):
            raise ValueError("duration_ms must be a finite number")
        duration_ms = float(duration_ms)
        expected_duration_ms = (
            SCOPE_SAMPLE_COUNT * sample_period_us / 1000.0
        )
        if not math.isfinite(duration_ms) or not math.isclose(
            duration_ms,
            expected_duration_ms,
            rel_tol=1e-6,
            abs_tol=0.05,
        ):
            raise ValueError(
                f"duration_ms must be {expected_duration_ms:g}"
            )
        if isinstance(pretrigger_ratio, bool) or not isinstance(
            pretrigger_ratio, (int, float)
        ):
            raise ValueError("pretrigger_ratio must be between 0.0 and 0.9")
        pretrigger_ratio = float(pretrigger_ratio)
        if (
            not math.isfinite(pretrigger_ratio)
            or pretrigger_ratio < 0.0
            or pretrigger_ratio > 0.9
        ):
            raise ValueError("pretrigger_ratio must be between 0.0 and 0.9")
        return duration_ms, pretrigger_ratio

    def download(
        self,
        *,
        decimation,
        sample_period_us,
        duration_ms,
        pretrigger_ratio,
        timeout=15.0,
    ):
        """Request, validate, decode, and chronologically rotate a capture."""

        duration_ms, pretrigger_ratio = self._validate_capture_metadata(
            decimation,
            sample_period_us,
            duration_ms,
            pretrigger_ratio,
        )
        if isinstance(timeout, bool) or not isinstance(timeout, (int, float)):
            raise ValueError("timeout must be a positive finite number")
        timeout = float(timeout)
        if not math.isfinite(timeout) or timeout <= 0:
            raise ValueError("timeout must be a positive finite number")

        self._write_command(b"D")
        deadline = time.monotonic() + timeout

        for _ in range(16):
            line = self._readline(deadline)
            if line == _BEGIN_RECORD:
                break
            if line.startswith("SCOPE-DATA/1 ERROR"):
                raise ScopeSerialError(line)
        else:
            raise ScopeSerialError("scope record start marker was not received")

        channel_line = self._readline(deadline)
        if not channel_line.startswith("#"):
            raise ScopeSerialError("scope channel header is malformed")
        channel_names = tuple(
            name for name in channel_line[1:].split(",") if name
        )
        if channel_names != SCOPE_CHANNEL_NAMES:
            raise ScopeSerialError(
                "scope channel header mismatch: "
                f"expected {SCOPE_CHANNEL_NAMES!r}, got {channel_names!r}"
            )

        final_index_line = self._readline(deadline)
        match = _FINAL_INDEX_RE.fullmatch(final_index_line)
        if match is None:
            raise ScopeSerialError("scope final-index line is malformed")
        final_index = int(match.group(1))
        if not 0 <= final_index < SCOPE_SAMPLE_COUNT:
            raise ScopeSerialError(
                f"scope final index {final_index} is out of range"
            )

        values = []
        value_count = SCOPE_SAMPLE_COUNT * len(SCOPE_CHANNEL_NAMES)
        for index in range(value_count):
            line = self._readline(deadline, maximum_length=16)
            if line == _END_RECORD:
                raise ScopeSerialError(
                    f"scope payload is truncated at {index}/{value_count} values"
                )
            if _HEX_VALUE_RE.fullmatch(line) is None:
                raise ScopeSerialError(
                    f"scope payload value {index} is not 8-digit hexadecimal"
                )
            values.append(struct.unpack(">f", bytes.fromhex(line))[0])

        terminator = self._readline(deadline, maximum_length=16)
        if terminator != _END_RECORD:
            if _HEX_VALUE_RE.fullmatch(terminator):
                raise ScopeSerialError(
                    f"scope payload contains more than {value_count} values"
                )
            raise ScopeSerialError("scope record end marker is malformed")

        rows = tuple(
            tuple(values[offset : offset + len(SCOPE_CHANNEL_NAMES)])
            for offset in range(0, value_count, len(SCOPE_CHANNEL_NAMES))
        )
        first_index = (final_index + 1) % SCOPE_SAMPLE_COUNT
        chronological_samples = rows[first_index:] + rows[:first_index]

        sample_period_s = sample_period_us * 1e-6
        trigger_offset = pretrigger_ratio * SCOPE_SAMPLE_COUNT
        time_axis_s = tuple(
            (index - trigger_offset) * sample_period_s
            for index in range(SCOPE_SAMPLE_COUNT)
        )

        return ScopeCapture(
            decimation=decimation,
            sample_period_us=sample_period_us,
            duration_ms=duration_ms,
            channel_names=channel_names,
            samples=chronological_samples,
            final_index=final_index,
            pretrigger_ratio=pretrigger_ratio,
            time_axis_s=time_axis_s,
        )

File: src/tools/ScopeSerial.m
%
% Copyright (c) 2021-present LAAS-CNRS
%
% SPDX-License-Identifier: GPL-2.0-or-later
%

classdef ScopeSerial < handle
    % ScopeSerial  Bounded parser for the dedicated scope-data interface.

    properties (SetAccess = private)
        Port
    end

    properties (Access = private)
        Serial
    end

    properties (Constant)
        BaudRate = 115200
        SampleCount = 1024
        ChannelNames = [ ...
            "V1Low_V", "V2Low_V", "VHigh_V", "I1Low_A", ...
            "I2Low_A", "IHigh_A", "Duty1", "Duty2"]
    end

    methods
        function obj = ScopeSerial(port, options)
            arguments
                port (1,1) string
                options.Timeout (1,1) double = 1.0
                options.ProbeTimeout (1,1) double = 3.0
                options.Transport = []
            end
            if strlength(port) == 0
                error("ScopeSerial:validation", ...
                    "scope port must be an explicit non-empty string");
            end
            ScopeSerial.validatePositiveFinite("Timeout", options.Timeout);
            ScopeSerial.validatePositiveFinite( ...
                "ProbeTimeout", options.ProbeTimeout);

            obj.Port = port;
            try
                if isempty(options.Transport)
                    obj.Serial = serialport( ...
                        port, ScopeSerial.BaudRate, ...
                        "Timeout", options.Timeout);
                else
                    if ~ismethod(options.Transport, "read") || ...
                            ~ismethod(options.Transport, "write")
                        error("ScopeSerial:validation", ...
                            "Transport must provide read and write methods");
                    end
                    obj.Serial = options.Transport;
                end
                obj.probe(options.ProbeTimeout);
            catch ME
                obj.close();
                if startsWith(string(ME.identifier), "ScopeSerial:")
                    rethrow(ME);
                end
                error("ScopeSerial:transport", ...
                    "failed to open scope data port %s: %s", ...
                    port, ME.message);
            end
        end

        function close(obj)
            if ~isempty(obj.Serial) && ismethod(obj.Serial, "close")
                obj.Serial.close();
            end
            obj.Serial = [];
        end

        function delete(obj)
            obj.close();
        end

        function capture = download(obj, metadata, options)
            % Request, validate, decode, and rotate a frozen capture.
            arguments
                obj
                metadata (1,1) struct
                options.Timeout (1,1) double = 15.0
            end
            ScopeSerial.validatePositiveFinite("Timeout", options.Timeout);
            metadata = ScopeSerial.validateMetadata(metadata);

            obj.writeCommand(uint8('D'));
            started = tic;
            foundStart = false;
            for i = 1:16
                line = obj.readLine(started, options.Timeout, 256);
                if line == "begin record"
                    foundStart = true;
                    break
                end
                if startsWith(line, "SCOPE-DATA/1 ERROR")
                    error("ScopeSerial:protocol", "%s", line);
                end
            end
            if ~foundStart
                error("ScopeSerial:protocol", ...
                    "scope record start marker was not received");
            end

            channelLine = obj.readLine(started, options.Timeout, 256);
            if ~startsWith(channelLine, "#")
                error("ScopeSerial:protocol", ...
                    "scope channel header is malformed");
            end
            channelNames = split(extractAfter(channelLine, 1), ",").';
            channelNames(channelNames == "") = [];
            if ~isequal(channelNames, ScopeSerial.ChannelNames)
                error("ScopeSerial:protocol", ...
                    "scope channel header mismatch");
            end

            finalLine = obj.readLine(started, options.Timeout, 256);
            token = regexp(finalLine, '^# ([0-9]+)$', 'tokens', 'once');
            if isempty(token)
                error("ScopeSerial:protocol", ...
                    "scope final-index line is malformed");
            end
            finalIndex = str2double(token{1});
            if finalIndex < 0 || finalIndex >= ScopeSerial.SampleCount
                error("ScopeSerial:protocol", ...
                    "scope final index is out of range");
            end

            channelCount = numel(ScopeSerial.ChannelNames);
            valueCount = ScopeSerial.SampleCount * channelCount;
            values = zeros(valueCount, 1);
            for i = 1:valueCount
                line = obj.readLine(started, options.Timeout, 16);
                if line == "end record"
                    error("ScopeSerial:protocol", ...
                        "scope payload is truncated at %d/%d values", ...
                        i - 1, valueCount);
                end
                if isempty(regexp(line, '^[0-9A-Fa-f]{8}$', 'once'))
                    error("ScopeSerial:protocol", ...
                        "scope payload value %d is not 8-digit hexadecimal", ...
                        i - 1);
                end
                bits = uint32(hex2dec(char(line)));
                values(i) = double(typecast(bits, 'single'));
            end

            terminator = obj.readLine(started, options.Timeout, 16);
            if terminator ~= "end record"
                if ~isempty(regexp(terminator, ...
                        '^[0-9A-Fa-f]{8}$', 'once'))
                    error("ScopeSerial:protocol", ...
                        "scope payload contains more than %d values", ...
                        valueCount);
                end
                error("ScopeSerial:protocol", ...
                    "scope record end marker is malformed");
            end

            rows = reshape(values, channelCount, ScopeSerial.SampleCount).';
            firstIndex = mod(finalIndex + 1, ScopeSerial.SampleCount);
            order = [ ...
                (firstIndex + 1):ScopeSerial.SampleCount, ...
                1:firstIndex];
            samples = rows(order, :);

            triggerOffset = ...
                metadata.pretriggerRatio * ScopeSerial.SampleCount;
            timeAxisS = ( ...
                (0:(ScopeSerial.SampleCount - 1)).' - triggerOffset) * ...
                metadata.samplePeriodUs * 1e-6;

            capture = struct( ...
                "decimation", metadata.decimation, ...
                "samplePeriodUs", metadata.samplePeriodUs, ...
                "durationMs", metadata.durationMs, ...
                "channelNames", channelNames, ...
                "samples", samples, ...
                "finalIndex", finalIndex, ...
                "pretriggerRatio", metadata.pretriggerRatio, ...
                "timeAxisS", timeAxisS);
        end
    end

    methods (Access = private)
        function writeCommand(obj, command)
            try
                if ismethod(obj.Serial, "flush")
                    flush(obj.Serial, "input");
                end
                write(obj.Serial, command, "uint8");
            catch ME
                if startsWith(string(ME.identifier), "ScopeSerial:")
                    rethrow(ME);
                end
                error("ScopeSerial:transport", ...
                    "scope data-port write failed: %s", ME.message);
            end
        end

        function line = readLine(obj, started, timeout, maximumLength)
            bytes = uint8.empty(1, 0);
            while toc(started) < timeout
                try
                    if obj.Serial.NumBytesAvailable == 0
                        pause(0.001);
                        continue
                    end
                    value = read(obj.Serial, 1, "uint8");
                catch ME
                    error("ScopeSerial:transport", ...
                        "scope data-port read failed: %s", ME.message);
                end
                if isempty(value)
                    continue
                end
                bytes(end+1) = uint8(value); %#ok<AGROW>
                if numel(bytes) > maximumLength
                    error("ScopeSerial:protocol", ...
                        "scope line exceeds %d bytes", maximumLength);
                end
                if bytes(end) == 10
                    while ~isempty(bytes) && ...
                            any(bytes(end) == uint8([10, 13]))
                        bytes(end) = [];
                    end
                    if any(bytes > 127)
                        error("ScopeSerial:protocol", ...
                            "scope response is not ASCII");
                    end
                    line = string(char(bytes));
                    return
                end
            end
            error("ScopeSerial:timeout", ...
                "scope data-port response timed out");
        end

        function probe(obj, timeout)
            obj.writeCommand(uint8('?'));
            started = tic;
            for i = 1:32
                line = obj.readLine(started, timeout, 256);
                if line == "SCOPE-DATA/1 OK"
                    return
                end
                if startsWith(line, "SCOPE-DATA/1 ERROR")
                    error("ScopeSerial:probe", ...
                        "scope probe rejected: %s", line);
                end
            end
            error("ScopeSerial:probe", ...
                "scope probe response was not received");
        end
    end

    methods (Static, Access = private)
        function validatePositiveFinite(name, value)
            if ~isnumeric(value) || ~isscalar(value) || ~isreal(value) || ...
                    ~isfinite(value) || value <= 0
                error("ScopeSerial:validation", ...
                    "%s must be a positive finite number", name);
            end
        end

        function metadata = validateMetadata(metadata)
            required = [ ...
                "decimation", "samplePeriodUs", ...
                "durationMs", "pretriggerRatio"];
            if ~all(isfield(metadata, cellstr(required)))
                error("ScopeSerial:validation", ...
                    "capture metadata is incomplete");
            end

            decimation = metadata.decimation;
            if ~isnumeric(decimation) || islogical(decimation) || ...
                    ~isscalar(decimation) || ~isreal(decimation) || ...
                    ~isfinite(decimation) || fix(decimation) ~= decimation || ...
                    decimation < 1 || decimation > 100
                error("ScopeSerial:validation", ...
                    "decimation must be an integer between 1 and 100");
            end
            samplePeriodUs = metadata.samplePeriodUs;
            if ~isnumeric(samplePeriodUs) || islogical(samplePeriodUs) || ...
                    ~isscalar(samplePeriodUs) || ...
                    samplePeriodUs ~= 100 * decimation
                error("ScopeSerial:validation", ...
                    "samplePeriodUs must equal 100 times decimation");
            end
            durationMs = metadata.durationMs;
            expectedDuration = ...
                ScopeSerial.SampleCount * samplePeriodUs / 1000;
            if ~isnumeric(durationMs) || islogical(durationMs) || ...
                    ~isscalar(durationMs) || ~isreal(durationMs) || ...
                    ~isfinite(durationMs) || ...
                    abs(durationMs - expectedDuration) > 0.05
                error("ScopeSerial:validation", ...
                    "durationMs does not match the capture timing");
            end
            pretriggerRatio = metadata.pretriggerRatio;
            if ~isnumeric(pretriggerRatio) || islogical(pretriggerRatio) || ...
                    ~isscalar(pretriggerRatio) || ...
                    ~isreal(pretriggerRatio) || ...
                    ~isfinite(pretriggerRatio) || ...
                    pretriggerRatio < 0 || pretriggerRatio > 0.9
                error("ScopeSerial:validation", ...
                    "pretriggerRatio must be between 0.0 and 0.9");
            end

            metadata = struct( ...
                "decimation", double(decimation), ...
                "samplePeriodUs", double(samplePeriodUs), ...
                "durationMs", double(durationMs), ...
                "pretriggerRatio", double(pretriggerRatio));
        end
    end
end
